### PR TITLE
Redshift Type and Dedupe

### PR DESCRIPTION
**Pull request by evantahler**
Typing and Deduping for Redshift.  It works, but it is kind of janky.  See the comments below as to why 😓

### CODE REVIEW

File: one-table.redshift.sql [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+-- DROP FUNCTION _airbyte_safe_cast_to_integer(input text)
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_integer(input text)
+RETURNS INTEGER
+STABLE AS $$
+	try:
+		v = int(input)
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
```
Comment (evantahler) [inline one-table.redshift.sql:54]: ✅ This feels rather safe and robust

File: one-table.redshift.sql [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+-- DROP FUNCTION _airbyte_safe_cast_to_boolean(input text)
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_boolean(input text)
+RETURNS BOOLEAN
+STABLE AS $$
+	try:
+		v = input.lower() == 'True' or input.lower() == 'true'
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
```
Comment (evantahler) [inline one-table.redshift.sql]: 🟡 we need to test this.  I think the Airbyte Protocol sends booleans as literal `true` and `false`, but I'm worried about the SQL<-->python translation of the values, which is why I added the 'lower'

File: one-table.redshift.sql [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+-- DROP FUNCTION _airbyte_safe_cast_to_boolean(input text)
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_boolean(input text)
+RETURNS BOOLEAN
+STABLE AS $$
+	try:
+		v = input.lower() == 'true'
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
```
Comment (evantahler) [inline one-table.redshift.sql:65]: ✅ I'm a little worried about the SQL<-->python bridge here, but lower-casing the string seems safe enough

File: one-table.redshift.sql [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+-- DROP FUNCTION _airbyte_safe_cast_to_text(input text);
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_text(input text)
+RETURNS TEXT
+STABLE AS $$
+	try:
+		v = str(input)
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
```
Comment (evantahler) [inline one-table.redshift.sql:76]: ✅ This feels rather safe and robust

File: one-table.redshift.sql [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+    , "id" int8 -- PK cannot be null, but after raw insert and before typing, row will be null
+    , "first_name" text
+    , "age" int8
+    , "address" super
+    , "updated_at" timestamp
+);
+
+---------------------------------------
+--------- SAFE CAST METHODS -----------
+---------------------------------------
```
[the remark this answers is not in the document]
Comment (evantahler) [reply one-table.redshift.sql:43]: What's also super annoying is that you can't pass a SUPER type into a function in Redshift, so we need to manually cast everything to `text` as part of the call

`super is not a supported parameter type for functions or procedures`

File: one-table.redshift.sql [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+--------- UTILS -----------
+---------------------------
+
+-- http://datasamurai.blogspot.com/2016/06/creating-uuid-function-in-redshift.html
+CREATE OR REPLACE FUNCTION gen_uuid()
+RETURNS VARCHAR(36)
+VOLATILE AS $$
+ import uuid
+ return uuid.uuid4().__str__()
+$$ LANGUAGE plpythonu;
```
Comment (evantahler) [inline one-table.redshift.sql:100]: There's no native way to generate a UUID in Redshift... so... we'll use 🐍

File: one-table.redshift.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (evantahler) [reply one-table.redshift.sql:43]: Also, remember that in python spaces and tabs matter...

File: one-table.redshift.sql [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+  missing_pk_count := (
+    SELECT COUNT(1)
+    FROM x.USERS_RAW
+    WHERE
+      "_airbyte_loaded_at" IS NULL
+      AND _airbyte_safe_cast_to_integer(_airbyte_data.id::text) IS NULL
+    );
+
+  IF missing_pk_count > 0 THEN
+    RAISE EXCEPTION 'Table has % rows missing a primary key', missing_pk_count;
```
Comment (michel-tricot) [inline one-table.redshift.sql:139]: does it mean the whole load will fail if one recodrd doesn't have a PK?

File: one-table.redshift.sql [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+    _airbyte_raw_id,
+    _airbyte_extracted_at
+  )
+
+  WITH intermediate_data AS (
+    SELECT
+       _airbyte_safe_cast_to_integer(_airbyte_data.id::text) as id
+      ,_airbyte_safe_cast_to_text(_airbyte_data.first_name::text) as first_name
+      ,_airbyte_safe_cast_to_integer(_airbyte_data.age::text) as age
+      ,_airbyte_data.address as address -- there is no way to make a JSON/SUPER safe cast in Redshift
```
Comment (michel-tricot) [inline one-table.redshift.sql:162]: what does it mean? Address will be a JSON?

Comment (evantahler) [reply one-table.redshift.sql:162]: Yep! That's what we want, and in redshift lingo "any sub-section of a SUPER-type remains a SUPER", so it's going to be valid... we just can't confirm that behavior.  Valid JSON in this case also includes strings, numbers, arrays, etc.

File: one-table.redshift.sql (hunk 7 above)
Comment (evantahler) [reply one-table.redshift.sql:139]: Yep (and a missing cursor will fail syncs as well)!
If you don't have a PK or cursor, we can't reliably tell you anything about the sync - which row had a problem, or which row should win a dedupe.

Comment (evantahler) [reply one-table.redshift.sql:139]: Any other column can be null

**Comment by evantahler** [comment]
In https://github.com/airbytehq/typing-and-deduping-sql/pull/24/commits/2c9680a2364f6829dcb505c59585f290133b8944 I added `_airbyte_safe_cast_to_number()` to test how decimals would, work and ensure that we maintain precision.  We are good, because we can use python's `decimal()` parser, rather than float.

Screenshot 2023-08-02 at 3 10 38 PM

**Pull request merged**